File: 2262-solving-questions-with-brainpower/2262-solving-questions-with-brainpower.cpp
#include <algorithm>

class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {


        int n = questions.size();
        vector<long long> dp(questions.size(), 0);

        

        dp[n-1] = questions[n-1][0];

        for (int i = n-2; i>= 0; i--){

            long long cur_q = questions[i][0];

            int next = questions[i][1] + i  + 1;


            if (next< questions.size()){
               cur_q += dp[next];
        }
        dp[i] = max(cur_q, dp[i+1]);
        }

        return *max_element(dp.begin(), dp.end());

    }
};

File: 2262-solving-questions-with-brainpower/2262-solving-questions-with-brainpower.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        

        n = len(questions)
        dp = [0] * n

        dp[-1] = questions[-1][0]

        i = n-2
        while i >= 0:

            if (i + questions[i][1]+1) < n:
                dp[i] = max(dp[i+1], dp[i+questions[i][1] +1] + questions[i][0])
            else:
                dp[i] = max(questions[i][0], dp[i+1])
                
            i-=1
            # print(dp)
        
        return max(dp)